### DISCUSSION TITLE
Float enum query parameter fails to validate but works in plain pydantic

### DISCUSSION THREAD

**Discussion by estan** [fastapi/fastapi#8858] [Questions] [answered]
### Example

```python
from enum import Enum

from fastapi import FastAPI
from pydantic import BaseModel


app = FastAPI()


class Number(float, Enum):
    ten = 10.0
    twelve = 12.0


class MyModel(BaseModel):
    num: Number


MyModel(num=10.0)  # <- This works!


@app.get("/")
async def root(num: Number):
    return {"message": f"Hello {num}"}
```

```
$ uvicorn test:app --reload
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [176218] using statreload
INFO:     Started server process [176220]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
```

```
$ curl -s http://localhost:8000?num=10.0
{"detail":[{"loc":["query","num"],"msg":"value is not a valid enumeration member; permitted: 10.0, 12.0","type":"type_error.enum","ctx":{"enum_values":[10.0,12.0]}}]}
```

### Description

* Open the browser and call the endpoint `/?num=10.0`.
* It returns a JSON with `{"detail":[{"loc":["query","num"],"msg":"value is not a valid enumeration member; permitted: 10.0, 12.0","type":"type_error.enum","ctx":{"enum_values":[10.0,12.0]}}]}`.
* But I expected it to return `{"message": "Hello 10.0"}`.

Note how `MyModel(num=10.0)` in the example above does not throw a `ValidationError`, so `pydantic` seems to validate this fine - it is only when calling the `/` FastAPI endpoint with `10.0` passed for the `num: Number` parameter that a validation error is returned.

### Environment

* OS: Linux
* FastAPI Version: 0.61.1
* Python version: 3.8.5

**Comment by Dustyposa** [comment]
That's all arguments is string in query parameter.
So this will raised an error:
```python
MyModel(num='10.0')
```

**Comment by estan** [comment]
Aha, I think I was confused since I found with `IntEnum` it works:

```python
from enum import IntEnum

from pydantic import BaseModel


class Number(IntEnum):
    ten = 10


class MyModel(BaseModel):
    num: Number


MyModel(num="10")
```

So I thought it would work with `class Number(float, Enum)` similarly, since `IntEnum` is just `class IntEnum(int, Enum)` in the standard library.

But I tried now and with `class Number(int, Enum)` it does not work:

```python
from enum import Enum

from pydantic import BaseModel


class Number(int, Enum):
    ten = 10


class MyModel(BaseModel):
    num: Number


MyModel(num="10")
```

This is despite the fact that `IntEnum` is defined in the standard library as just `class IntEnum(int, Enum)`.

So I guess pydantic must have some special code in it specifically for the type `IntEnum`, and that's why that works with `IntEnum`(?)

Also, `int("10")` and `float("10.0")` both work, so it would be convenient if the framework could support float enums.

What do I do if I want an endpoint to take and validate a parameter with an allowed list of values that are not strings (e.g. floats)? Must I take a general `float` parameter and do the validation manually?

At one point in the FastAPI [docs](https://fastapi.tiangolo.com/tutorial/path-params/#get-the-enumeration-value) regarding enums,  it says:

> You can get the actual value (**a str in this case**) using model_name.value, or in general, your_enum_member.value:

Note the "a str in this case", suggesting that there are other cases, i.e. that you could use enums with some other type. But I guess `IntEnum` is the only workable option?

**Comment by Hultner** [comment]
I'd expect this to work, from my point of view I'd regard it as a bug. In swagger it shows the number without the decimal as well.

Not sure if there's a simple way to add custom parsers for incoming data to FastAPI (couldn't find anything obvious in the docs). But you should be able to solve it with a custom middleware. 

I had a similar problem with Decimals and fixed it in pydantic a while back but it was the other way around. Decimals without any decimal places would add a trailing .0 and always fail the request.

**Comment by estan** [comment]
Alright. In any case, this is a limitation in pydantic, so I guess I should close this and file an issue there (?).

**Comment by Hultner** [comment]
> Alright. In any case, this is a limitation in pydantic, so I guess I should close this and file an issue there (?).

I'm not completely sure if it is since you can actually solve this for pydantic models quite easily by adding a custom encoder to the config. It's hard for pydantic to cover all possible coercion's, especially with custom classes like this. It would make sense if I could somehow specify a custom decoder for this particular type. I took a quick peak into the FastAPI source code but I'm not familiar enough with how this is handled to be completely sure where to look.

You could of-course also solve it by using a custom function or class for you query params which takes these in as a float but create the enum as it's initiated.

**Comment by Hultner** [comment, marked as answer]
Here's an example how you could solve your problem. 

```python
from enum import Enum

from fastapi import FastAPI, Depends, HTTPException
from pydantic import BaseModel


app = FastAPI()


class Number(float, Enum):
    ten = 10.0
    twelve = 12.0


async def number_parameter(num: float) -> Number:
    try:
        return Number(num)
    except ValueError as err:
        raise HTTPException(detail=err.args[0], status_code=422)


class MyModel(BaseModel):
    num: Number


MyModel(num=10.0)  # <- This works!


@app.get("/")
async def root(num: Number = Depends(number_parameter)):
    return {"message": f"Hello {num}"}

```

The downside with this approach is that you aren't getting the correct OpenAPI specs, it shows up as a float instead. This could probably be achieved via the extra args on Query. And I create a slightly more simplistic error.

Examples of how it works:

```
http ":8000?num=12.0"
HTTP/1.1 200 OK
content-length: 24
content-type: application/json
date: Wed, 19 May 2021 11:58:13 GMT
server: uvicorn

{
    "message": "Hello 12.0"
}

```


```
http ":8000?num=13.0"
HTTP/1.1 422 Unprocessable Entity
content-length: 39
content-type: application/json
date: Wed, 19 May 2021 11:58:09 GMT
server: uvicorn

{
    "detail": "13.0 is not a valid Number"
}

```

**Comment by Hultner** [comment]
You could also improve your errors a bit by passing them from pydantic instead piggybacking the model you created like this.
I also patch the schema using the definitions from `MyModel.schema()["definitions"]["Number"]` with this you're fully set with a correct schema as well.

```python
from enum import Enum

from fastapi import FastAPI, Depends, HTTPException, Query
from pydantic import BaseModel, ValidationError


app = FastAPI()


class Number(float, Enum):
    ten = 10.0
    twelve = 12.0


class MyModel(BaseModel):
    num: Number


async def number_parameter(
    num: float = Query(..., **MyModel.schema()["definitions"]["Number"])
) -> Number:
    try:
        tmp_num = MyModel(num=num)
        return tmp_num.num
    except ValidationError as err:
        # breakpoint()
        raise HTTPException(detail=err.errors(), status_code=422)


# MyModel(num=10.0)  # <- This works!


@app.get("/")
async def root(num: Number = Depends(number_parameter)):
    return {"message": f"Hello {num}"}
```

Gives errors likes this
```
http ":8000?num=13.0"
HTTP/1.1 400 Bad Request
content-length: 158
content-type: application/json
date: Wed, 19 May 2021 12:16:03 GMT
server: uvicorn

{
    "detail": [
        {
            "ctx": {
                "enum_values": [
                    10.0,
                    12.0
                ]
            },
            "loc": [
                "num"
            ],
            "msg": "value is not a valid enumeration member; permitted: 10.0, 12.0",
            "type": "type_error.enum"
        }
    ]
}
```

And a schema like this:
```json
http :8000/openapi.json
HTTP/1.1 200 OK
content-length: 1031
content-type: application/json
date: Wed, 19 May 2021 12:34:44 GMT
server: uvicorn

{
    "components": {
        "schemas": {
            "HTTPValidationError": {
                "properties": {
                    "detail": {
                        "items": {
                            "$ref": "#/components/schemas/ValidationError"
                        },
                        "title": "Detail",
                        "type": "array"
                    }
                },
                "title": "HTTPValidationError",
                "type": "object"
            },
            "ValidationError": {
                "properties": {
                    "loc": {
                        "items": {
                            "type": "string"
                        },
                        "title": "Location",
                        "type": "array"
                    },
                    "msg": {
                        "title": "Message",
                        "type": "string"
                    },
                    "type": {
                        "title": "Error Type",
                        "type": "string"
                    }
                },
                "required": [
                    "loc",
                    "msg",
                    "type"
                ],
                "title": "ValidationError",
                "type": "object"
            }
        }
    },
    "info": {
        "title": "FastAPI",
        "version": "0.1.0"
    },
    "openapi": "3.0.2",
    "paths": {
        "/": {
            "get": {
                "operationId": "root__get",
                "parameters": [
                    {
                        "description": "An enumeration.",
                        "in": "query",
                        "name": "num",
                        "required": true,
                        "schema": {
                            "description": "An enumeration.",
                            "enum": [
                                10.0,
                                12.0
                            ],
                            "title": "Number",
                            "type": "number"
                        }
                    }
                ],
                "responses": {
                    "200": {
                        "content": {
                            "application/json": {
                                "schema": {}
                            }
                        },
                        "description": "Successful Response"
                    },
                    "422": {
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/HTTPValidationError"
                                }
                            }
                        },
                        "description": "Validation Error"
                    }
                },
                "summary": "Root"
            }
        }
    }
}

```

Edit: Added schema definitions as well.

**Comment by estan** [comment]
Wow. Let me digest this in a while.

**Comment by Hultner** [comment]
> Wow. Let me digest this in a while.

Yes, hope it solves your problems. Tell me if you want me to elaborate on anything.
On a side note, nice to see a fellow Swedish FastAPI user :)
